autoprotocol/protocol.py

```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
                 self.instructions[-1].to.append(xfer)
             else:
                 provision_instructions_to_return.append(
-                    self._append_and_return(Provision(resource_id, dest_group))
+                    self._append_and_return(
+                        Provision(resource_id, dest_group, measurement_mode)
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
+        unique_measure_modes = set()
+        for amount in volumes:
+            print(amount.dimensionality)
+            if not isinstance(amount, (str, Unit)):
```

Comment:
this check is a bit unwieldy. 

First, we should not be allowing `amount` to be of type `str` as the subsequent steps call an attribute that only `Unit` objects have.

Second, we could use something like `amount = parse_unit(amount)` to sanely parse the unit provided. It will guarantee `amount` is type `Unit`

Reply:
I see below code is redundant:
```
            if not isinstance(amount, (str, Unit)):
                raise TypeError(f"Provided amount {amount} is not supported.")
```
because it is handled by below code which is before `_identify_measurement_mode ` call:
```
        if not isinstance(provision_amounts, list):
            provision_amounts = [Unit(provision_amounts)] * len(dests)
        else:
            if len(provision_amounts) != len(dests):
                raise RuntimeError(
                    "To provision a resource into multiple "
                    "destinations with multiple volumes or masses, the  "
                    "list of volumes or masses must correspond with the "
                    "destinations in length and in order."
                )
            provision_amounts = [Unit(v) for v in provision_amounts]
```

Reply:
Ah I see. Its already been cast by the time it reaches here. 

In that case, I think its fine to remove the `str` type-check. I.e. just ` if not isinstance(amount, Unit)`

Reply:
Yep